pkg/infrastructure/vsphere/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,426 @@
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/govc/importx"
+	"github.com/vmware/govmomi/nfc"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/ovf"
+	"github.com/vmware/govmomi/vapi/tags"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+	"sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/session"
+
+	"github.com/openshift/installer/pkg/asset"
+	vcentercontexts "github.com/openshift/installer/pkg/asset/cluster/vsphere"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/rhcos"
+	"github.com/openshift/installer/pkg/infrastructure/clusterapi"
+	"github.com/openshift/installer/pkg/rhcos/cache"
+	"github.com/openshift/installer/pkg/types/vsphere"
+)
+
+// Provider is the vSphere implementation of the clusterapi InfraProvider.
+type Provider struct {
+	clusterapi.InfraProvider
+}
+
+var _ clusterapi.PreProvider = Provider{}
+
+// Name returns the vsphere provider name.
+func (p Provider) Name() string {
+	return vsphere.Name
+}
+
+func attachTag(ctx context.Context, session *session.Session, vmMoRefValue, tagID string) error {
+	tagManager := session.TagManager
+
+	moRef := types.ManagedObjectReference{
+		Value: vmMoRefValue,
+		Type:  "VirtualMachine",
+	}
+
+	err := tagManager.AttachTag(ctx, tagID, moRef)
+
+	if err != nil {
+		return fmt.Errorf("unable to attach tag: %w", err)
+	}
+	return nil
+}
+
+func findAvailableHostSystems(ctx context.Context, session *session.Session, clusterHostSystems []*object.HostSystem) (*object.HostSystem, error) {
+	var hostSystemManagedObject mo.HostSystem
+	for _, hostObj := range clusterHostSystems {
+		err := hostObj.Properties(ctx, hostObj.Reference(), []string{"config.product", "network", "datastore", "runtime"}, &hostSystemManagedObject)
+		if err != nil {
+			return nil, err
+		}
+		if hostSystemManagedObject.Runtime.InMaintenanceMode {
+			continue
+		}
+		return hostObj, nil
+	}
+	return nil, errors.New("all hosts unavailable")
+}
+
+// Used govc/importx/ovf.go as an example to implement
+// resourceVspherePrivateImportOvaCreate and upload functions
+// See: https://github.com/vmware/govmomi/blob/cc10a0758d5b4d4873388bcea417251d1ad03e42/govc/importx/ovf.go#L196-L324
+func upload(ctx context.Context, archive *importx.ArchiveFlag, lease *nfc.Lease, item nfc.FileItem) error {
+	file := item.Path
+
+	f, size, err := archive.Open(file)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	opts := soap.Upload{
+		ContentLength: size,
+	}
+
+	return lease.Upload(ctx, item, f, opts)
+}
+
+func importRhcosOva(ctx context.Context, session *session.Session, folder *object.Folder, cachedImage, clusterID, tagID, diskProvisioningType string, failureDomain vsphere.FailureDomain) error {
```

Comment:
This code might be a bit easier to follow by breaking these functions into their own files

Reply:
ack, will work on that